src/experimental/utils.cpp

```diff
@@ -0,0 +1,242 @@
+                * resolution [kt_double]: Cell resolution
+             * Return:
+                * Void
+             */
+            if (final_grid_pos.GetX() < robot_grid_pos.GetX() && final_grid_pos.GetY() >= robot_grid_pos.GetY())
+            {
+                final_x[0] = limit_x + resolution;
+                final_x[2] = limit_x + resolution;
+
+                cell_limits[2] = limit_y;
+                cell_limits[3] = limit_y + resolution;
+            }
+
+            if (final_grid_pos.GetX() >= robot_grid_pos.GetX() && final_grid_pos.GetY() < robot_grid_pos.GetY())
+            {
+                initial_y[2] = limit_y - resolution;
+                initial_y[3] = limit_y - resolution;
+
+                final_y[1] = limit_y - resolution;
+                final_y[3] = limit_y - resolution;
+
+                cell_limits[2] = limit_y - resolution;
+                cell_limits[3] = limit_y;
+            }
+        }
+
+
+        int signum(int num)
+        {
+            /**
+             * Get the sign of an operation, used by Bresenham algorithm
+             * Arguments:
+                * num [int]: Number for perform the sign operation
+             * Return:
+                * int: Sign
+             */
+            if (num < 0) return -1;
+            if (num >= 1) return 1;
+            return 0;
+        }
+
+
+        void clearVisitedCells(Eigen::MatrixXd & grid)
+        {
+            /**
+             * Clear the given floating Eigen::Matrix
+             * Arguments:
+                * grid [Eigen::Matrix]: Grid for cleaning
+            * Return:
+                * Void
+             */
+            grid.setZero();
+        }
+
+
+        void clearVisitedCells(Eigen::MatrixXi & grid)
+        {
+            /**
+             * Clear the given integer Eigen::Matrix
+             * Arguments:
+                * grid [Eigen::Matrix]: Grid for cleaning
+             * Return:
+                * Void
+             */
+
+            grid.setZero();
+        }
+
+
+        karto::Vector2<int> getGridPosition(karto::Vector2<kt_double> const& position, kt_double resolution)
+        {
+            /**
+             * Mapping a continuous position into a grid position
+             * Arguments:
+                * pose [karto::Vector2<kt_double>]: Continuos pose
+                * resolution [kt_double]: Cell resolution
+             * Return:
+                * karto::Vector2<int>: Grid position
+             */
+            int x_cell = floor((position.GetX() / resolution));
+            int y_cell = floor((position.GetY() / resolution));
+
+            return karto::Vector2<int>{x_cell, y_cell};
+        }
+
+
+        karto::Vector2<kt_double> calculateCellIntersectionPoints(karto::Vector2<kt_double> const & laser_start,
+            karto::Vector2<kt_double> const & laser_end, karto::Vector2<kt_double> const & cell_start, karto::Vector2<kt_double> const & cell_end)
```

Comment:
Thanks for this suggestion. 

From my perspective it is only worth the implementation of `Segment2`, because this kind of data structure is used a lot within this code. It was defined as a template structure in `utils.hpp` as shown next:

```C++
    template<typename T>
    struct Segment2
    {
        karto::Vector2<T> start;
        karto::Vector2<T> end;
    };
```
Regarding the Box2 I decided to not implement that structure for two reasons.
1. It is used only a couple of times, to pass the cell corners to find the intersection points between a laser beam and the current cell.
2. Sometimes and depending on the where the cell is located WRO the laser beam origin we need to modify the corners. So I found that it might be confusing to handle the changes with the `Box2` structure. Please consider the following structure definition (Maybe the naming convention is what seems confusing for me):

```C++
    template<typename T>
    struct Box2
    {
        karto::Vector2<T> bl; // bottom left
        karto::Vector2<T> br; // bottom right
        karto::Vector2<T> tl; // top left
        karto::Vector2<T> tr; // top right
    };
```
I will leave this thread open!

Reply:
@kmilo7204 hmm, I don't follow why `Box2` definition requires coordinates for each corner. Unless It's a four sided polygon, coordinates for two corners along a diagonal should always suffice.